vm.c

```diff
@@ -2968,20 +2977,67 @@ free_loading_table_entry(st_data_t key, st_data_t value, st_data_t arg)
     return ST_DELETE;
 }
 
+void rb_free_loaded_features_index(rb_vm_t *vm);
+void rb_objspace_free_objects(void *objspace);
+
 int
 ruby_vm_destruct(rb_vm_t *vm)
 {
     RUBY_FREE_ENTER("vm");
 
     if (vm) {
         rb_thread_t *th = vm->ractor.main_thread;
-        struct rb_objspace *objspace = vm->objspace;
-        vm->ractor.main_thread = NULL;
+        VALUE *stack = th->ec->vm_stack;
+        if (rb_free_on_shutdown) {
+            rb_free_default_rand_key();
+            rb_free_encoded_insn_data();
+            rb_free_global_enc_table();
+            rb_free_loaded_builtin_table();
+
+            rb_free_shared_fiber_pool();
+            rb_free_static_symid_str();
+            rb_free_transcoder_table();
+            rb_free_vm_opt_tables();
+            rb_free_warning();
+            rb_free_rb_global_tbl();
+            rb_free_loaded_features_index(vm);
+            rb_ractor_t *r = vm->ractor.main_ractor;
+            xfree(r->sync.recv_queue.baskets);
+            xfree(r->sync.takers_queue.baskets);
+
+            rb_id_table_free(vm->negative_cme_table);
+            st_free_table(vm->overloaded_cme_table);
+
+            rb_id_table_free(RCLASS(rb_mRubyVMFrozenCore)->m_tbl);
+
+            rb_shape_t *cursor = rb_shape_get_root_shape();
+            rb_shape_t *end = rb_shape_get_shape_by_id(GET_SHAPE_TREE()->next_shape_id);
+            while (cursor < end) {
+                // 0x1 == SINGLE_CHILD_P
+                if (cursor->edges && !(((uintptr_t)cursor->edges) & 0x1))
+                    rb_id_table_free(cursor->edges);
+                cursor += 1;
+            }
+
+            xfree(GET_SHAPE_TREE());
+
+            st_free_table(vm->static_ext_inits);
+            st_free_table(vm->ensure_rollback_table);
+
+            ruby_xfree(vm->postponed_job_buffer);
```

Comment:
While working on another issue with the postponed job infrastructure, I realised there's something a little bit annoying about doing this... It's the most miniscule of corner cases, but I figure I'll bring it up.

If some other gem has registered a signal handler, it's possible a signal might arrive after this and before the process exits, the gem's signal signal handler could then try and call `rb_register_postponed_job`, which will be a use-after-free here.

Stackprof in particular _tries_ to get this right, and almost does: it registers a `ruby_vm_at_exit` hook here: https://github.com/tmm1/stackprof/blob/91d12e77b3d3c5014870b4ee7c19eab24ad04b5b/ext/stackprof/stackprof.c#L903 to disable itself.

There are just a couple teeny tiny problems though:

The first problem is, that the free-on-shutdown stuff is happening before `ruby_vm_run_at_exit_hooks`. Easy fix, let's hoist that further up I suppose? Or push this stuff further down?

The second problem is what happens in the presence of non-ruby-managed threads. If there are other native threads in the process, a signal handler could be in-progress on that thread already when `ruby_vm_run_at_exit_hooks` is called, so setting `ruby_vm_running = 0;` wouldn't stop it from happening.

I actually can't think of anything at all we can do about that though. If there was a way to run something at exit after all _other_ threads have been destroyed, we could do that, but there isn't such a thing. I guess well-behaved extensions need to make sure they `pthread_join` all their threads in a `ruby_vm_at_exit` handler?